File: ios/Classes/Mapping/Symbology/SymbolFactory.swift
//
// Created by Valentin Grigorean on 18.04.2024.
//

import Foundation
import ArcGIS

class SymbolFactory {
    static func createSymbol(data: Any?) -> Symbol? {
        guard let data = data as? Dictionary<String, Any> else {
            return nil
        }

        guard let type = data["type"] as? String else {
            return nil
        }
        switch type {
        case "simple-marker":
            let simpleMarker = SimpleMarkerSymbol()
            simpleMarker.interpretSimpleMarkerSymbol(data: data)
            return simpleMarker
        case "simple-line":
            let simpleLine = SimpleLineSymbol()
            simpleLine.interpretLineSymbol(data: data)
            return simpleLine
        case "simple-fill":
            let simpleFill = SimpleFillSymbol()
            simpleFill.interpretSimpleFillSymbol(data: data)
            return simpleFill
        case "picture-marker":
            return createPictureMarkerSymbol(data: data)
        case "text":
            let textSymbol = TextSymbol()
            textSymbol.interpretTextSymbol(data: data)
            return textSymbol
        default:
            return nil
        }
    }

    private static func createPictureMarkerSymbol(data: Dictionary<String, Any>) -> PictureMarkerSymbol {
        var symbol: PictureMarkerSymbol
        if let url = data["url"] as? String {
            symbol = PictureMarkerSymbol(url: URL(string: url)!)
        } else if let resource = data["resource"] as? String {
            let tintColor = UIColor(data: data["tintColor"])
            var image = UIImage(named: resource)!
            if let color = tintColor {
                image = image.colored(color)
            }
            symbol = PictureMarkerSymbol(image: image)
        } else if let fromBytes = data["fromBytes"] as? FlutterStandardTypedData {
            symbol = PictureMarkerSymbol(image: UIImage(data: fromBytes.data)!)
        } else {
            symbol = PictureMarkerSymbol()
        }

        symbol.interpretPictureMarkerSymbol(data: data)
        return symbol
    }
}

File: android/src/main/kotlin/com/valentingrigorean/arcgis_maps_flutter/mapping/symbology/SymbolFactory.kt
package com.valentingrigorean.arcgis_maps_flutter.mapping.symbology

import android.content.Context
import com.arcgismaps.mapping.symbology.CompositeSymbol
import com.arcgismaps.mapping.symbology.PictureMarkerSymbol
import com.arcgismaps.mapping.symbology.SimpleFillSymbol
import com.arcgismaps.mapping.symbology.SimpleLineSymbol
import com.arcgismaps.mapping.symbology.SimpleMarkerSymbol
import com.arcgismaps.mapping.symbology.Symbol
import com.arcgismaps.mapping.symbology.TextSymbol
import com.valentingrigorean.arcgis_maps_flutter.convert.mapping.symbology.interpretPictureMarkerSymbol
import com.valentingrigorean.arcgis_maps_flutter.convert.mapping.symbology.interpretSimpleFillSymbol
import com.valentingrigorean.arcgis_maps_flutter.convert.mapping.symbology.interpretSimpleLineSymbol
import com.valentingrigorean.arcgis_maps_flutter.convert.mapping.symbology.interpretSimpleMarkerSymbol
import com.valentingrigorean.arcgis_maps_flutter.convert.mapping.symbology.interpretTextSymbol
import com.valentingrigorean.arcgis_maps_flutter.convert.toBitmapDrawable
import com.valentingrigorean.arcgis_maps_flutter.utils.createDrawableBitmap

object SymbolFactory {
    fun createSymbol(context: Context, data: Any?): Symbol? {
        if (data !is Map<*, *>) {
            return null
        }
        return when (data["type"] as String) {
            "json" -> Symbol.fromJsonOrNull(data["data"] as String)
            "simple-line" -> SimpleLineSymbol().apply {
                interpretSimpleLineSymbol(data)
            }

            "simple-fill" -> SimpleFillSymbol().apply {
                interpretSimpleFillSymbol(data)
            }

            "simple-marker" -> SimpleMarkerSymbol().apply {
                interpretSimpleMarkerSymbol(data)
            }

            "picture-marker" -> createPictureMarkerSymbol(context, data)
            "text" -> TextSymbol().apply {
                interpretTextSymbol(data)
            }

            "composite" -> CompositeSymbol().apply {
                data["symbols"]?.let {
                    val symbolsRaw = it as List<*>
                    symbols.addAll(symbolsRaw.mapNotNull { symbolRaw ->
                        createSymbol(context, symbolRaw)
                    })
                }
            }

            else -> null
        }
    }


    private fun createPictureMarkerSymbol(context: Context, data: Map<*, *>): PictureMarkerSymbol {
        val symbol: PictureMarkerSymbol = if (data.containsKey("url")) {
            PictureMarkerSymbol(data["url"] as String)
        } else if (data.containsKey("resource")) {
            val resourceName = data["resource"] as String
            val tintColor = data["tintColor"] as Long?
            val bitmap = context.createDrawableBitmap(resourceName, tintColor?.toInt())
            bitmap?.let { PictureMarkerSymbol.createWithImage(it) } ?: PictureMarkerSymbol()
        } else if (data.containsKey("fromBytes")) {
            val bytes = data["fromBytes"] as ByteArray
            PictureMarkerSymbol.createWithImage(bytes.toBitmapDrawable(context))
        } else {
            PictureMarkerSymbol()
        }
        symbol.interpretPictureMarkerSymbol(data)
        return symbol
    }
}